ligo/stablecoin/spec.ligo

```diff
@@ -0,0 +1,257 @@
+
+type operator_transfer_policy is
+| Operator_transfer_permitted
+| Operator_transfer_denied
+
+type owner_transfer_policy is
+| Owner_no_op
+| Optional_owner_hook
+| Required_owner_hook
+
+type custom_permission_policy is record
+  tag       : string
+; config_api : option (address)
+end
+
+type permissions_descriptor is record
+  self     : self_transfer_policy
+; operator : operator_transfer_policy
+; receiver : owner_transfer_policy
+; sender   : owner_transfer_policy
+; custom   : option (custom_permission_policy)
+end
+
+type permissions_descriptor_params is
+  contract (permissions_descriptor)
+
+type pause_params is unit
+
+type unpause_params is unit
+
+type configure_minter_params is address
+
+type set_minting_allowance_params is record
+  minter : address
+; amount : nat
+end
+
+type remove_minter_params is address
+
+type mint_params is record
+  to_    : address
+; amount : nat
+end
+
+type burn_params is record
+  from_  : address
+; amount : nat
+end
+
+type parameter is
+  // FA2
+  Transfer                 of transfer_params
+| Balance_of                of balance_of_params
+| Total_supply           of total_supply_params
+| Get_token_metadata         of get_token_metadata_params
+| Permissions_descriptor of permissions_descriptor_params
+| Update_operators          of update_operator_params
+| Is_operator               of is_operator_params
+  // Custom
+| Pause               of pause_params
+| Unpause             of unpause_params
+| Configure_minter     of configure_minter_params
+| Set_minting_allowance of set_minting_allowance_params
+| Remove_minter        of remove_minter_params
+| Mint                of mint_params
+| Burn                of burn_params
+
+/* ------------------------------------------------------------- */
+
+/*
+ * Hooks
+ */
+
+type registry is set (address)
+
+// TODO: currently it's the same as `transfer` but with optional
+// participants.
+type transfer_descriptor is record
+  from_    : option (address)
+; to_      : option (address)
+; token_id : token_id
+; amount   : nat
+end
+
+type transfer_descriptor_param is record
+  contract_address : address
+; batch    : list (transfer_descriptor)
+; operator : address
+end
+
+type token_receiver is
+| Tokens_received of transfer_descriptor_param
+
+type token_sender is
+| Tokens_sent of transfer_descriptor_param
+
+type set_hook_param is record
+  hook : unit -> contract (transfer_descriptor_param)
+; permissions_descriptor : permissions_descriptor
+end
+
+// Entrypoints with param hook
+type closed_parameter is
+| Parameter of parameter
+| Set_transfer_hook of set_hook_param
+
+/* ------------------------------------------------------------- */
+
+/*
+ * Storage
+ */
+
+type operator_actions is
+| All_operator_tokens
+| Some_operator_tokens of set (token_id)
+| All_operator_tokens_except of set (token_id)
+
+type owner is address
+type operator is address
+
+type operator_storage is
+  big_map ((owner * operator), operator_actions)
```

Comment:
It's useless to store `operator_actions` value in case of a single token contract. An operator can operate only one type of token, that's it.